pkg/pool-weighted/contracts/managed/ManagedPoolSettings.sol

```diff
@@ -755,52 +724,88 @@ abstract contract ManagedPoolSettings is BasePool, ProtocolFeeCache, ReentrancyG
+        // they would include the removed token.
         _ensureNoWeightChange();
 
-        // Total supply is potentially changing so we collect AUM fees.
+        // Total supply is potentially changing so we collect AUM fees. For consistency, we do this unconditionally.
         _collectAumManagementFees(supply);
 
         // Before this function is called, the caller must have withdrawn all balance for `token` from the Pool. This
         // means that the Pool is in an invalid state, since among other things the invariant is zero. Because we're not
         // in a valid state and all value-changing operations will revert, we are free to modify the Pool state (e.g.
         // alter weights).
+        //
         // We don't need to test the zero balance since the Vault will simply revert on deregistration if this is not
-        // the case.
+        // the case, or if the token is not currently registered.
+        PoolRegistrationLib.deregisterToken(getVault(), getPoolId(), tokenToRemove);
 
-        // Removing a token will cause for the weights of all other tokens to increase. This is fine, as there is no
-        // maximum weight. We also don't need to check that the new token exists in the Pool, as the Vault will simply
-        // revert if we try to deregister a token that is not registered.
-        // We do, however, want to check that the Pool will end up with at least two tokens. This simplifies some
-        // assumptions made elsewhere (e.g. the denormalized weight sum will always be non-zero), and doesn't greatly
-        // restrict the controller.
+        (IERC20[] memory tokens, , ) = getVault().getPoolTokens(getPoolId());
+        _require(tokens.length >= 2, Errors.MIN_TOKENS);
 
-        (IERC20[] memory tokens, ) = _getPoolTokens();
-        _require(tokens.length > 2, Errors.MIN_TOKENS);
+        // Once we've updated the state in the Vault, we need to also update our own state. This is a two-step process,
+        // since we need to:
+        //  a) delete the state of the remoevd token
+        //  b) adjust the weights of all other tokens
 
-        uint256 tokenNormalizedWeight = ManagedPoolTokenLib.getTokenWeight(
-            _tokenState[token],
-            ManagedPoolStorageLib.getGradualWeightChangeProgress(_poolState),
-            _denormWeightSum
+        // Deleting the old token is straightforward. The Pool itself doesn't track how many or which tokens it uses
+        // (and relies instead on the Vault for this), so we simply delete the token-specific information. We first read
+        // its weight however, since we'll need it later.
+        uint256 weightChangeProgress = ManagedPoolStorageLib.getGradualWeightChangeProgress(_poolState);
+        uint256 tokenToRemoveWeight = ManagedPoolTokenLib.getTokenWeight(
+            _tokenState[tokenToRemove],
+            weightChangeProgress
         );
```

Comment:
```suggestion
        // We've ensured that the most recent weight change is complete.
        uint256 tokenToRemoveWeight = ManagedPoolTokenLib.getTokenWeight(
            _tokenState[tokenToRemove],
            FixedPoint.ONE
        );
```

If you want to optimizoor.

Reply:
Ah interesting. `_getNormalizedWeights` will read the progress again though :)